### PR TITLE
eventhandler: minor refactor of task_handler

### PR DESCRIPTION
**Pull request by adnxn**
### Summary

eventhandler: minor refactor of task_handler
### Implementation details

1. removed taskHandler global in task_handler.go
2. HandleEngineEvents in handler.go owns an instance of taskHandler
3. moved taskHandler related types to task_handler.go

### Testing

- [X] Builds on Linux (`make release`)
- [X] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [X] Unit tests on Linux (`make test`) pass
- [X] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [X] Integration tests on Linux (`make run-integ-tests`) pass
- [X] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [X] Functional tests on Linux (`make run-functional-tests`) pass
- [X] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: 
no new tests.
### Description for the changelog

none.

### Licensing

This contribution is under the terms of the Apache 2.0 License: 
yes

### CODE REVIEW

File: agent/eventhandler/handler.go [hunk 1]
```diff
@@ -28,6 +28,8 @@ var statesaver statemanager.Saver = statemanager.NewNoopStateManager()
 
 func HandleEngineEvents(taskEngine engine.TaskEngine, client api.ECSClient, saver statemanager.Saver) {
 	statesaver = saver
+	handler := NewTaskHandler()
```
Comment (samuelkarp) [inline agent/eventhandler/handler.go]: It looks like now you have two separate `taskHandler`s: one here and one in `agent/acs/handler/payload_handler.go`.  Is that intentional?

Comment (adnxn) [inline agent/eventhandler/handler.go]: it's intentional. there is a clear division of labor between the two and i could rename this to something more descriptive. maybe "taskEventHandler"?

Comment (aaithal) [inline agent/eventhandler/handler.go]: What's the division of labor between the two? It seems like we should have only one of these if we're looking to consolidate these in a single entity.

File: agent/eventhandler/task_handler.go [hunk 2]
```diff
@@ -15,63 +15,80 @@ package eventhandler
...
 
-var handler *taskHandler
+// Maximum number of tasks that may be handled at once by the taskHandler
+const concurrentEventCalls = 3
 
-func init() {
-	handler = newTaskHandler()
+type eventList struct {
+	sending    bool // whether the list is already being handled
+	sync.Mutex      // Locks both the list and sending bool
```
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: Can you make these named fields?

File: agent/eventhandler/task_handler.go [hunk 3]
```diff
@@ -15,63 +15,80 @@ package eventhandler
...
+	*list.List      // list of *sendableEvents
+}
+
+type taskHandler struct {
+	submitSemaphore utils.Semaphore       // Semaphore on the number of tasks that may be handled at once
+	taskMap         map[string]*eventList // arn:*eventList map so events may be serialized per task
+	sync.RWMutex                          // Lock for the taskMap
```
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: Can you make this a named field?

File: agent/eventhandler/task_handler.go [hunk 4]
```diff
@@ -15,63 +15,80 @@ package eventhandler
...
+}
+
+func NewTaskHandler() *taskHandler {
```
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: A global method returning a local data type doesn't seem like the right thing to do in terms of api abstraction and design. Please make `taskHandler` a global type as it has methods that are global as well.

File: agent/acs/handler/payload_handler.go [hunk 5]
```diff
@@ -47,7 +48,7 @@ type payloadRequestHandler struct {
 }
 
 // newPayloadRequestHandler returns a new payloadRequestHandler object
-func newPayloadRequestHandler(ctx context.Context, taskEngine engine.TaskEngine, ecsClient api.ECSClient, cluster string, containerInstanceArn string, acsClient wsclient.ClientServer, saver statemanager.Saver, refreshHandler refreshCredentialsHandler, credentialsManager credentials.Manager) payloadRequestHandler {
+func newPayloadRequestHandler(ctx context.Context, taskEngine engine.TaskEngine, ecsClient api.ECSClient, cluster string, containerInstanceArn string, acsClient wsclient.ClientServer, saver statemanager.Saver, refreshHandler refreshCredentialsHandler, credentialsManager credentials.Manager, taskHandler *eventhandler.TaskHandler) payloadRequestHandler {
```
Comment (aaithal) [inline agent/acs/handler/payload_handler.go]: Can you split this long line into multiple lines? Makes it more readable IMO.

File: agent/acs/handler/payload_handler_test.go [hunk 6]
```diff
@@ -45,9 +46,10 @@ func TestHandlePayloadMessageWithNoMessageId(t *testing.T) {
 	ecsClient := mock_api.NewMockECSClient(ctrl)
 	stateManager := statemanager.NewNoopStateManager()
 	credentialsManager := credentials.NewManager()
+	taskHandler := eventhandler.NewTaskHandler()
 
 	ctx := context.Background()
-	buffer := newPayloadRequestHandler(ctx, taskEngine, ecsClient, clusterName, containerInstanceArn, nil, stateManager, refreshCredentialsHandler{}, credentialsManager)
+	buffer := newPayloadRequestHandler(ctx, taskEngine, ecsClient, clusterName, containerInstanceArn, nil, stateManager, refreshCredentialsHandler{}, credentialsManager, taskHandler)
```
Comment (aaithal) [inline agent/acs/handler/payload_handler_test.go]: This too, might benefit from splitting into multiple lines. There are multiple such occurrences in the diff and you might want to change all of those.

File: agent/api/statechange.go [hunk 7]
```diff
@@ -52,7 +52,7 @@ type TaskStateChange struct {
 	Reason string
 
 	// Task is a pointer to the task involved in the state change that gives the event handler a hook into storing
-	// what status was sent.  This is used to ensure the same event is handled only once.
+	// what stats was sent.  This is used to ensure the same event is handled only once.
```
Comment (aaithal) [inline agent/api/statechange.go]: This is supposed to be `status`. Why did you change this?

File: agent/eventhandler/task_handler.go [hunk 8]
```diff
@@ -15,63 +15,80 @@ package eventhandler
...
-var handler *taskHandler
+// Maximum number of tasks that may be handled at once by the TaskHandler
+const concurrentEventCalls = 3
 
-func init() {
-	handler = newTaskHandler()
+type eventList struct {
+	*list.List            // list of *sendableEvents
+	sending    bool       // whether the list is already being handled
+	mu         sync.Mutex // Locks both the list and sending bool
```
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: Please use a more descriptive name for this: `lock` or better yet specify what this lock is supposed to protect. `listLock` for example.

File: agent/eventhandler/task_handler.go [hunk 9]
```diff
@@ -15,63 +15,80 @@ package eventhandler
...
 )
 
-var handler *taskHandler
+// Maximum number of tasks that may be handled at once by the TaskHandler
+const concurrentEventCalls = 3
 
-func init() {
-	handler = newTaskHandler()
+type eventList struct {
+	*list.List            // list of *sendableEvents
```
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: Can you make this a non-embedded field as well? Example: `events *list.List`

File: agent/eventhandler/task_handler.go [hunk 10]
```diff
@@ -15,63 +15,80 @@ package eventhandler
...
+	sending    bool       // whether the list is already being handled
+	mu         sync.Mutex // Locks both the list and sending bool
+}
+
+type TaskHandler struct {
+	submitSemaphore utils.Semaphore       // Semaphore on the number of tasks that may be handled at once
+	taskMap         map[string]*eventList // arn:*eventList map so events may be serialized per task
```
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: `tasksToEvents` is a better name for this.

File: agent/eventhandler/task_handler.go [hunk 11]
```diff
@@ -15,63 +15,80 @@ package eventhandler
...
+	taskMapMu       sync.RWMutex          // Lock for the taskMap
```
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: If you modify the previous line as per my suggestion, `tasksToEventsLock` is a better name for this. In any case, let's change it from `taskMapMu` to something else, preferably with the phrase `lock` somewhere in the mix.

File: agent/eventhandler/task_handler.go [hunk 12]
```diff
@@ -15,63 +15,80 @@ package eventhandler
...
+}
+
+func NewTaskHandler() *TaskHandler {
+	taskMap := make(map[string]*eventList)
+	submitSemaphore := utils.NewSemaphore(concurrentEventCalls)
+
+	return &TaskHandler{
```
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: You can just do this and reduce loc's:

```
return &TaskHandler{
 taskMap: make(map[string]*eventList),
 submitSemaphore: utils.NewSemaphore(concurrentEventCalls),
}
```

File: agent/eventhandler/task_handler.go [hunk 13]
```diff
@@ -15,63 +15,80 @@ package eventhandler
...
-		if !preexisting {
-			log.Debug("New event", "change", change)
-			taskList = &eventList{List: list.New(), sending: false}
-			handler.taskMap[change.taskArn()] = taskList
-		}
-	}()
+	events, ok := handler.taskMap[change.taskArn()]
+
+	if !ok {
+		log.Debug("New event", "change", change)
```
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: While we are refactoring, can you also to using `seelog` instead of this logger?

File: agent/eventhandler/task_handler.go [hunk 14]
```diff
@@ -15,63 +15,80 @@ package eventhandler
...
+func (handler *TaskHandler) addEvent(change *sendableEvent, client api.ECSClient) {
+
 	log.Info("Adding event", "change", change)
 
-	// TaskEvents lock scope
-	func() {
-		handler.Lock()
-		defer handler.Unlock()
+	handler.taskMapMu.Lock()
+	defer handler.taskMapMu.Unlock()
```
Comment (jahkeup) [inline agent/eventhandler/task_handler.go]: This `defer`'ed `handler.taskMapMu.Unlock()` should be moved at the end of the operation on the map to prevent contention while waiting on the rest of the method instead of `defer`-ing to the end of the function scope.

Comment (jahkeup) [inline agent/eventhandler/task_handler.go]: In fact this [section](https://github.com/adnxn/amazon-ecs-agent/blob/74d261ebe34c28ac63987ccd3335e6f86c736ee7/agent/eventhandler/task_handler.go#L66-L75) could be moved to a function like `handler.getTaskEventList(taskArn string) (list *eventList)` to hold the lock in a smaller and descriptive scope.

File: agent/api/statechange.go (hunk 7 above)
Comment (adnxn) [inline agent/api/statechange.go]: oops, this is a typo.

File: agent/eventhandler/task_handler.go [hunk 15]
```diff
@@ -15,63 +15,96 @@ package eventhandler
...
-		}
-	}()
-
-	taskList.Lock()
-	defer taskList.Unlock()
+func (handler *TaskHandler) addEvent(change *sendableEvent, client api.ECSClient) {
+
+	seelog.Info("TaskHandler, Adding event: ", change)
+
+	eventsToSubmit := handler.getTaskEventList(change)
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: I think `taskEvents` would be more clear here than `eventsToSubmit`.

File: agent/eventhandler/task_handler.go [hunk 16]
```diff
@@ -15,63 +15,96 @@ package eventhandler
...
+	if !eventsToSubmit.sending {
+		eventsToSubmit.sending = true
+		go handler.SubmitTaskEvents(eventsToSubmit, client)
 	}
 }
 
+// getTaskEventList gets the eventList from taskToEvent map, and reduces the
+// scope of the taskToEventsLock to just this function
+func (handler *TaskHandler) getTaskEventList(change *sendableEvent) (eventsToSubmit *eventList) {
+
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: nit: remove blank line

File: agent/eventhandler/task_handler.go [hunk 17]
```diff
@@ -15,63 +15,96 @@ package eventhandler
...
+	handler.tasksToEventsLock.Lock()
+	defer handler.tasksToEventsLock.Unlock()
+
+	eventsToSubmit, ok := handler.tasksToEvents[change.taskArn()]
+
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: nit: remove blank line

File: agent/eventhandler/task_handler.go [hunk 18]
```diff
@@ -15,63 +15,96 @@ package eventhandler
...
+	if !ok {
+		seelog.Debug("TaskHandler, New event: ", change)
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: I know you copied the line from the old code, but "new event" really isn't super clear.  This is an event arriving for a task that does not currently have any events queued; all the events are "new".  Maybe "collecting events for new task"?

File: agent/eventhandler/task_handler.go [hunk 19]
```diff
@@ -15,63 +15,96 @@ package eventhandler
...
+	eventListLock sync.Mutex
+}
+
+// TaskHandler encapsulates the the map of a task arn to task and container events
+// associated with said task
+type TaskHandler struct {
+	// submitSemaphore for the number of tasks that may be handled at once
+	submitSemaphore utils.Semaphore
+	// taskToEvents is arn:*eventList map so events may be serialized per task
+	tasksToEvents map[string]*eventList
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: I know this is moved code, but I never spotted it there before: this is a memory leak because nothing ever gets removed from this map.  Can you add a TODO here to address that leak?

File: agent/eventhandler/task_handler.go [hunk 20]
```diff
@@ -15,63 +15,96 @@ package eventhandler
...
 	"time"
 
 	"github.com/aws/amazon-ecs-agent/agent/api"
 	"github.com/aws/amazon-ecs-agent/agent/utils"
+	"github.com/cihub/seelog"
 )
 
-var handler *taskHandler
+// Maximum number of tasks that may be handled at once by the TaskHandler
+const concurrentEventCalls = 3
```
Comment (samuelkarp) [inline agent/eventhandler/task_handler.go]: It's worth marking this down to revisit, especially since you're looking to continue modifying event submission.

File: agent/acs/handler/payload_handler.go [hunk 21]
```diff
@@ -274,7 +286,7 @@ func (payloadHandler *payloadRequestHandler) handleUnrecognizedTask(task *ecsacs
 	}
 
 	// Only need to stop the task; it brings down the containers too.
-	eventhandler.AddTaskEvent(api.TaskStateChange{
+	payloadHandler.taskHandler.AddTaskEvent(api.TaskStateChange{
```
Comment (richardpen) [inline agent/acs/handler/payload_handler.go]: Is this the only reason that `taskHandler` needs to be part of the `payloadHandler`? We may implement this by using the `taskEngine.TaskEvents()`, since `taskEngine` is already part of the `payloadHandler` so that `payloadHandler` can get rid of the `taskHandler`?

File: agent/eventhandler/task_handler.go [hunk 22]
```diff
@@ -15,63 +15,96 @@ package eventhandler
...
-			log.Debug("New event", "change", change)
-			taskList = &eventList{List: list.New(), sending: false}
-			handler.taskMap[change.taskArn()] = taskList
-		}
-	}()
-
-	taskList.Lock()
-	defer taskList.Unlock()
+func (handler *TaskHandler) addEvent(change *sendableEvent, client api.ECSClient) {
+
```
Comment (richardpen) [inline agent/eventhandler/task_handler.go]: nit: remove the blank line.

Comment (adnxn) [inline agent/eventhandler/task_handler.go]: fixed.

File: agent/acs/handler/payload_handler.go (hunk 21 above)
Comment (adnxn) [inline agent/acs/handler/payload_handler.go]: yes this is currently the only use, although i'm not that reimplementing the all the event submission logic would be a good idea. @aaithal any comments?

File: agent/eventhandler/task_handler.go (hunk 20 above)
Comment (adnxn) [inline agent/eventhandler/task_handler.go]: noted.

File: agent/eventhandler/task_handler.go (hunk 15 above)
Comment (adnxn) [inline agent/eventhandler/task_handler.go]: fair enough. that makes sense. fixed.

File: agent/eventhandler/task_handler.go (hunk 19 above)
Comment (adnxn) [inline agent/eventhandler/task_handler.go]: noted.

File: agent/eventhandler/task_handler.go (hunk 16 above)
Comment (adnxn) [inline agent/eventhandler/task_handler.go]: fixed.

File: agent/eventhandler/task_handler.go (hunk 18 above)
Comment (adnxn) [inline agent/eventhandler/task_handler.go]: you're right, "new event" doesn't really add much context here. fixed.

File: agent/eventhandler/task_handler.go (hunk 17 above)
Comment (adnxn) [inline agent/eventhandler/task_handler.go]: fixed.

File: agent/eventhandler/task_handler.go (hunk 20 above)
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: @adnxn Can you create an issue for this so that we don't lose track of it?

File: agent/acs/handler/payload_handler.go (hunk 21 above)
Comment (aaithal) [inline agent/acs/handler/payload_handler.go]: I'd prefer to leave it as is. The reasoning is that `engine.taskEvents` is being used to generate events exclusively from task engine and using that to pipe events from `payloadhandler` to the event handler seems like leaking the abstraction.

File: agent/eventhandler/task_handler.go [hunk 23]
```diff
@@ -15,63 +15,94 @@ package eventhandler
...
+	eventListLock sync.Mutex
+}
+
+// TaskHandler encapsulates the the map of a task arn to task and container events
+// associated with said task
+type TaskHandler struct {
+	// submitSemaphore for the number of tasks that may be handled at once
+	submitSemaphore utils.Semaphore
+	// taskToEvents is arn:*eventList map so events may be serialized per task
+	//TODO: fix leak, currently items are never removed from this map
```
Comment (aaithal) [inline agent/eventhandler/task_handler.go]: Can you create an issue to track this as well?

File: agent/eventhandler/task_handler.go (hunk 20 above)
Comment (adnxn) [inline agent/eventhandler/task_handler.go]: added: https://github.com/aws/amazon-ecs-agent/issues/797

File: agent/eventhandler/task_handler.go (hunk 23 above)
Comment (adnxn) [inline agent/eventhandler/task_handler.go]: added: https://github.com/aws/amazon-ecs-agent/issues/798

File: agent/acs/handler/payload_handler.go (hunk 21 above)
Comment (adnxn) [inline agent/acs/handler/payload_handler.go]: @aaithal  do you mind clarifying? what exactly do you mean by "seems like leaking the abstraction"? thanks =]

Comment (aaithal) [inline agent/acs/handler/payload_handler.go]: It's mostly what I mentioned earlier. Payload handler needs to generate an event, which is processed by the event handler. Task Engine has no role in this interaction. Just because Task Engine can also send events to the event handler doesn't mean that, that should be the interface used for sending events to the event stream. Does that make sense?

Comment (adnxn) [inline agent/acs/handler/payload_handler.go]: ah ok, makes sense. thanks.
